### PR TITLE
feat: Garbage collector

### LINKED ISSUE #145: Implement a Garbage Collector (GC) controller
The primary responsibility of the GC controller would be to watch the OADP/Velero install NS and clean up orphaned Velero Objects (VB, VR and VBSL) i.e. which do not have corresponding NA-Objects in the cluster.

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

~Blocked by https://github.com/migtools/oadp-non-admin/pull/139~

~Blocked by https://github.com/openshift/oadp-operator/pull/1623~

Fix #145 

If for some reason a non admin object is deleted without waiting its finalizer to delete the related Velero object, that object will become a leftover in OADP namespace, which non admin user can not clean up. To avoid these scenarios, garbage collector will run periodically to clean these possible leftover objects.

## How to test the changes made

Follow [install-from-source](https://github.com/migtools/oadp-non-admin/blob/master/docs/CONTRIBUTING.md#install-from-source) testing documentation, pointing this branch and my OADP fork branch [`feat/nac-garbage-collector`](https://github.com/mateusoliveira43/oadp-operator/tree/feat/nac-garbage-collector)

- Create a DPA with non admin enabled
- Create some NABs and NARs
- Delete DPA
- Remove finalizers from NABs and NARs and delete them
- Leftover related Backups and Restores will remain in OADP namespace
- Recreate DPA
- Leftover related Backups and Restores in OADP namespace should be deleted

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -166,6 +167,16 @@ func main() {
 		os.Exit(1)
 	}
 	// +kubebuilder:scaffold:builder
+	if err = (&controller.GarbageCollectorReconciler{
+		Client:        mgr.GetClient(),
+		Scheme:        mgr.GetScheme(),
+		OADPNamespace: oadpNamespace,
+		// TODO user input
```
Comment (mateusoliveira43) [inline cmd/main.go:174]: TODO DPA PR

File: internal/controller/garbagecollector_controller.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		if err := r.List(ctx, veleroBackupList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+			logger.Error(err, "Unable to fetch Backups in OADP namespace")
+			return ctrl.Result{}, err
+		}
+		for _, backup := range veleroBackupList.Items {
+			annotations := backup.GetAnnotations()
+			// TODO check UUID as well?
+			if !function.CheckVeleroBackupAnnotations(&backup) {
+				logger.V(1).Info("Backup does not have required annotations", constant.NameString, backup.Name)
+				// TODO delete?
```
Comment (mateusoliveira43) [inline internal/controller/garbagecollector_controller.go:77]: object has NAC labels, but no NAC annotations. What should we do?

File: internal/controller/garbagecollector_controller.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				continue
+			}
+			nab := &nacv1alpha1.NonAdminBackup{}
+			err := r.Get(ctx, types.NamespacedName{
+				Name:      annotations[constant.NabOriginNameAnnotation],
+				Namespace: annotations[constant.NabOriginNamespaceAnnotation],
```
Comment (mateusoliveira43) [inline internal/controller/garbagecollector_controller.go:83]: TODO test with non existent namespace to see what is the error

File: internal/controller/garbagecollector_controller.go (hunk 2 above)
Comment (mpryc) [reply internal/controller/garbagecollector_controller.go:77]: This should not happen https://github.com/migtools/oadp-non-admin/blob/25e40d6abd8c49c43a1a89a9f49d6d489d07707e/internal/controller/nonadminbackup_controller.go#L638 but if for some reason it does I think it would be logged as warning or INFO not a DEBUG. Alternative if you want this to be visible in OpenShift’s Events tab, you can emit an Event from your controller. This will help cluster admins and users see what’s happening without digging into logs.

File: internal/controller/garbagecollector_controller.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+		constant.ManagedByLabel: constant.ManagedByLabelValue,
+	}
+
+	if previousGarbageCollectorRun == nil || time.Now().After(previousGarbageCollectorRun.Add(r.Frequency)) {
+		previousGarbageCollectorRun = ptr.To(time.Now())
+		logger.V(1).Info("Garbage Collector Reconcile start")
+		// TODO duplication in delete logic
+		// TODO do deletion in parallel?
+
+		// TODO delete secret as well?
```
Comment (mateusoliveira43) [inline internal/controller/garbagecollector_controller.go:64]: should also check for possible leftover secrets?

File: internal/controller/garbagecollector_controller.go (hunk 3 above)
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:112]: not found error as well

File: internal/controller/garbagecollector_controller.go (hunk 2 above)
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:106]: I think there is no warning level in this lib we use

what is the decision here?

File: internal/controller/garbagecollector_controller.go (hunk 4 above)
Comment (mpryc) [reply internal/controller/garbagecollector_controller.go:64]: Correct, those should be cleaned up by the NABSL controller. On top of that now when I think about that we may want to add OwnerReference between Secret <-> Velero BSL object. This will cause Secret to be removed if the Velero BSL object is gone, so we don't have to implement this part in GC coontroller ?

Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:64]: that would be good

but secret is created prior to bsl, can be added independent of the creation order?

File: internal/controller/garbagecollector_controller.go (hunk 2 above)
Comment (mpryc) [reply internal/controller/garbagecollector_controller.go:106]: Annotations are only to ensure ownership reference exists between NA* and Velero objects.
If labels are there it means that they are managed and created by the controller.
If annotations are missing it means that for some different update/patch to the Velero object the ownership reference was removed.

On a second thought I believe this is something that is going to be subject for GC removal.

Even if the Velero Syncs back objects from the S3 storage to the openshift-adp namespace, those annotations are preserved.

File: internal/controller/garbagecollector_controller.go (hunk 4 above)
Comment (mpryc) [reply internal/controller/garbagecollector_controller.go:64]: You are right. We can not add or patch the *ownerReference* on the Secret which already exists. What we could do is to add it vice-versa to the Velero BSL and make Velero BSL having *ownerReference* on the secret and then add `blockOwnerDeletion` on the Secret object. This should prevent secret to be removed independently from the BSL, but as always it's not 100% guarantee and there are edge cases where this still may happen.

**Comment by mpryc** [comment]
@mateusoliveira43 I am not sure if you will need this part, but implemented in case you will: https://github.com/migtools/oadp-non-admin/pull/162

**Comment by mpryc** [comment]
@mateusoliveira43 @shubham-pampattiwar After going again through our drawings I found that #162 isn't really what we discussed and implemented the https://github.com/migtools/oadp-non-admin/pull/163 which should now satisfy the deletion workflow and GC.

File: go.mod [hunk 5]
```diff
@@ -82,3 +82,5 @@ require (
 )
 
 replace github.com/vmware-tanzu/velero => github.com/openshift/velero v0.10.2-0.20241211163542-fa8f2486175b
+
+replace github.com/openshift/oadp-operator => github.com/mateusoliveira43/oadp-operator v0.0.0-20250131150824-b010f27c07bb
```
Comment (mateusoliveira43) [inline go.mod:86]: TODO remove prior to merge

File: internal/controller/garbagecollector_controller.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	// TODO do deletion in parallel?
+
+	secretList := &corev1.SecretList{}
+	if err := r.List(ctx, secretList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+		logger.Error(err, "Unable to fetch Secret in OADP namespace")
+		return ctrl.Result{}, err
+	}
+	for _, secret := range secretList.Items {
+		annotations := secret.GetAnnotations()
+		// TODO check UUID as well?
```
Comment (shubham-pampattiwar) [inline internal/controller/garbagecollector_controller.go:69]: As we are using UUID to establish relationship among objects, lets use it to confirm whether we want to classify the object as orphan/residual or not before deleting it.

File: internal/controller/garbagecollector_controller.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	veleroBackupStorageLocationList := &velerov1.BackupStorageLocationList{}
+	if err := r.List(ctx, veleroBackupStorageLocationList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+		logger.Error(err, "Unable to fetch BackupStorageLocations in OADP namespace")
+		return ctrl.Result{}, err
+	}
+	for _, backupStorageLocation := range veleroBackupStorageLocationList.Items {
+		annotations := backupStorageLocation.GetAnnotations()
+		// TODO check UUID as well?
```
Comment (shubham-pampattiwar) [inline internal/controller/garbagecollector_controller.go:100]: Similar comments as above for BSL, Backup and Restore objects.

File: cmd/main.go [hunk 8]
```diff
@@ -191,32 +204,42 @@ func main() {
...
-		return nil, nil, err
+		return dpaConfiguration, err
 	}
-	enforcedBackupSpec := &velerov1.BackupSpec{}
-	enforcedRestoreSpec := &velerov1.RestoreSpec{}
 	for _, dpa := range dpaList.Items {
 		if dpa.Namespace == oadpNamespace {
 			if dpa.Spec.NonAdmin != nil && dpa.Spec.NonAdmin.EnforceBackupSpec != nil {
-				enforcedBackupSpec = dpa.Spec.NonAdmin.EnforceBackupSpec
+				dpaConfiguration.EnforceBackupSpec = dpa.Spec.NonAdmin.EnforceBackupSpec
```
[the remark this answers is not in the document]
Comment (mateusoliveira43) [reply cmd/main.go:233]: will apply change in next commit

File: internal/controller/garbagecollector_controller.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	logger.V(1).Info("Garbage Collector Reconcile end")
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GarbageCollectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("garbagecollector").
```
[the remark this answers is not in the document]
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:193]: will apply change in next commit

File: internal/controller/garbagecollector_controller.go [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	if err := r.List(ctx, secretList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+		logger.Error(err, "Unable to fetch Secret in OADP namespace")
+		return ctrl.Result{}, err
+	}
+	for _, secret := range secretList.Items {
+		annotations := secret.GetAnnotations()
+		// TODO check UUID as well?
+		if !function.CheckVeleroBackupStorageLocationAnnotations(&secret) {
+			logger.V(1).Info("Secret does not have required annotations", constant.NameString, secret.Name)
+			// TODO delete?
```
[the remark this answers is not in the document]
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:72]: should we do this now or can be later improvement (before release of 1.5 but after current sprint)

[the remark this answers is not in the document]
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go:72]: this is question is for all 4 objects. this scenario object has non admin labels, but does not have non admin annotations. I suspect this could happen if something removed annotations from our created object

File: internal/controller/garbagecollector_controller.go (hunk 6 above)
Comment (mateusoliveira43) [reply internal/controller/garbagecollector_controller.go]: done

File: cmd/main.go [hunk 11]
```diff
@@ -191,32 +204,42 @@ func main() {
 	}
 }
 
-func getEnforcedSpec(restConfig *rest.Config, oadpNamespace string) (*velerov1.BackupSpec, *velerov1.RestoreSpec, error) {
+func getDPAConfiguration(restConfig *rest.Config, oadpNamespace string) (v1alpha1.NonAdmin, error) {
+	dpaConfiguration := v1alpha1.NonAdmin{
+		GarbageCollectionPeriod: &metav1.Duration{
+			Duration: 24 * time.Hour, //nolint:revive // 1 day
```
[the remark this answers is not in the document]
Comment (kaovilai) [reply cmd/main.go:210]: Up to you. Could disable that linter

[the remark this answers is not in the document]
Comment (shubham-pampattiwar) [reply cmd/main.go:210]: +1 @kaovilai 
@mateusoliveira43 I see one more place where the linter must be complaining https://github.com/migtools/oadp-non-admin/pull/152/files#diff-ce3b46308e9edc7774e29eb0fbc7d14f3a68144e1f1650ea9a6018fba22a8942R35

File: internal/controller/garbagecollector_controller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (shubham-pampattiwar) [reply internal/controller/garbagecollector_controller.go:76]: Then in such a scenario the user can decide whether to manually delete/cleanup the resource or not. Us deleting it could cause problems IMHO.

File: internal/controller/garbagecollector_controller.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+	logger := log.FromContext(ctx)
+
+	labelSelector := client.MatchingLabels{
+		constant.OadpLabel:      constant.OadpLabelValue,
+		constant.ManagedByLabel: constant.ManagedByLabelValue,
+	}
+
+	logger.V(1).Info("Garbage Collector Reconcile start")
+	// TODO duplication in delete logic
+	// TODO do deletion in parallel?
```
[the remark this answers is not in the document]
Comment (shubham-pampattiwar) [reply internal/controller/garbagecollector_controller.go]: yeah works, I created an issue to track it: https://github.com/migtools/oadp-non-admin/issues/167

**Review by shubham-pampattiwar — COMMENTED** [review]
worked for me:
```
2025-02-05T18:51:11Z    DEBUG   orphan Backup deleted   {"controller": "nonadmingarbagecollector", "reconcileID": "8da7ee0b-242d-43f3-a8be-fc8dedbfbf52", "name": "nacuser-ns-nab-obj-4a8ccbc9-bfe5-4f81-9032-64d3cb87b840"}
2025-02-05T18:51:11Z    DEBUG   Garbage Collector Reconcile end {"controller": "nonadmingarbagecollector", "reconcileID": "8da7ee0b-242d-43f3-a8be-fc8dedbfbf52"}
```
please resolve the comments and we are good to go on this one. Thank you !

**Review by shubham-pampattiwar — APPROVED** [review]

File: cmd/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaovilai) [reply cmd/main.go:210]: dupword and revive has to be one of the most annoying ones, who thought this was a good idea :D

**Review by kaovilai — APPROVED** [review]

**Review by kaovilai — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by weshayutin** [comment]
```
2025-02-07T18:13:42Z DEBUG Garbage Collector Reconcile start {"controller": "nonadmingarbagecollector", "reconcileID": "643c8bef-1ff8-47bf-a0e4-6adca38f842a"}
2025-02-07T18:13:42Z INFO Starting workers {"controller": "nonadminbackupstoragelocation", "controllerGroup": "oadp.openshift.io", "controllerKind": "NonAdminBackupStorageLocation", "worker count": 1}
2025-02-07T18:13:42Z INFO Starting workers {"controller": "nonadminbackup", "controllerGroup": "oadp.openshift.io", "controllerKind": "NonAdminBackup", "worker count": 1}
2025-02-07T18:13:42Z INFO Starting workers {"controller": "nonadminrestore", "controllerGroup": "oadp.openshift.io", "controllerKind": "NonAdminRestore", "worker count": 1}
2025-02-07T18:13:42Z DEBUG orphan Backup deleted {"controller": "nonadmingarbagecollector", "reconcileID": "643c8bef-1ff8-47bf-a0e4-6adca38f842a", "name": "nacuser1-nacuser1-backup-1-58151d1d-7234-4539-8bea-f48fed76f814"}
2025-02-07T18:13:42Z DEBUG orphan Backup deleted {"controller": "nonadmingarbagecollector", "reconcileID": "643c8bef-1ff8-47bf-a0e4-6adca38f842a", "name": "nacuser1-nacuser1-backup-2-4fddb883-ea2b-48da-8077-50d33e7cd150"}
2025-02-07T18:13:42Z DEBUG orphan Restore deleted {"controller": "nonadmingarbagecollector", "reconcileID": "643c8bef-1ff8-47bf-a0e4-6adca38f842a", "name": "nacuser1-nacuser1-restore--013ec860-4c1b-42c0-bfc5-e5065e50ddc5"}
2025-02-07T18:13:42Z DEBUG Rejected Restore Update event: no changes to the CompletionTimestamp in the VeleroRestore object {"VeleroRestoreQueuePredicate": {"name":"nacuser1-nacuser1-restore--013ec860-4c1b-42c0-bfc5-e5065e50ddc5","namespace":"openshift-adp"}}
```
One can see the deletion.. and I think they were deleted.  but they came back

```
oc get backup -n openshift-adp
NAME                                                              AGE
nacuser1-nacuser1-backup-1-58151d1d-7234-4539-8bea-f48fed76f814   6m40s
nacuser1-nacuser1-backup-2-4fddb883-ea2b-48da-8077-50d33e7cd150   6m40s
```

perhaps my cluster is not right.. will want to look at this w/ @mateusoliveira43 
perhaps the garbage collector and the sync controller were fighting?

**Comment by mateusoliveira43** [comment]
@weshayutin this is expected

you deleted with `oc delete` your NAB resource, right? this means backup files in object storage are still there. When velero sync controller runs, backup CR will be recreated in OADP namespace

**Comment by mateusoliveira43** [comment]
but one thing we can take a way from this test, is that does not make sense to non admin sync controller frequency (still not in the code) be greater than or equal to non admin garbage collection frequency, or will have scenario where NABs will never be synced to non admin namespaces. I will add this validation in non admin sync controller PR

**Comment by weshayutin** [comment]
> @weshayutin this is expected
> 
> you deleted with `oc delete` your NAB resource, right? this means backup files in object storage are still there. When velero sync controller runs, backup CR will be recreated in OADP namespace

yup..   agreed that is what happened.
I think we should stub out some test cases for this for @PrasadJoshi12 after we merge. So we're very clear on the silly little edge cases

**Review by weshayutin — APPROVED** [review]
/LGTM
